libopenage/pathfinding/integration_field.cpp

```diff
@@ -0,0 +1,520 @@
+// Copyright 2024-2024 the openage authors. See copying.md for legal info.
+
+#include "integration_field.h"
+
+#include <cmath>
+
+#include "error/error.h"
+#include "log/log.h"
+
+#include "coord/tile.h"
+#include "pathfinding/cost_field.h"
+#include "pathfinding/definitions.h"
+#include "pathfinding/portal.h"
+
+
+namespace openage::path {
+
+IntegrationField::IntegrationField(size_t size) :
+	size{size},
+	cells(this->size * this->size, INTEGRATE_INIT) {
+	log::log(DBG << "Created integration field with size " << this->size << "x" << this->size);
+}
+
+size_t IntegrationField::get_size() const {
+	return this->size;
+}
+
+const integrated_t &IntegrationField::get_cell(const coord::tile &pos) const {
+	return this->cells.at(pos.ne + pos.se * this->size);
+}
+
+const integrated_t &IntegrationField::get_cell(size_t idx) const {
+	return this->cells.at(idx);
+}
+
+std::vector<size_t> IntegrationField::integrate_los(const std::shared_ptr<CostField> &cost_field,
+                                                    const coord::tile &target) {
+	ENSURE(cost_field->get_size() == this->get_size(),
+	       "cost field size "
+	           << cost_field->get_size() << "x" << cost_field->get_size()
+	           << " does not match integration field size "
+	           << this->get_size() << "x" << this->get_size());
+
+	// Store the wavefront_blocked cells
+	std::vector<size_t> wavefront_blocked;
+
+	// Target cell index
+	auto target_idx = target.ne + target.se * this->size;
+
+	// Lookup table for cells that have been found
+	std::unordered_set<size_t> found;
+	found.reserve(this->size * this->size);
+
+	// Cells that still have to be visited by the current wave
+	std::deque<size_t> current_wave;
+
+	// Cells that have to be visited in the next wave
+	std::deque<size_t> next_wave;
+
+	// Cost of the current wave
+	integrated_cost_t cost = INTEGRATED_COST_START;
+
+	// Move outwards from the target cell, updating the integration field
+	current_wave.push_back(target_idx);
+	do {
```

Comment:
Technically it is a flood fill algorithm since we are just incrementing the cost after each wave.

The blocked lines are the boundaries of the area that is flooded. Remember that this method is supposed to determine which cells are in line of sight (LOS). The blocked lines are the edges of the LOS area.

Reply:
Basically this part: 
![grafik](https://github.com/user-attachments/assets/a991ab76-9282-490a-88ec-bec589c39f5d)